File: workshop/20190323-clean-architecture/py-sample/sample-04.py
import sys
from abc import *

class TaxEntity:
    __taxRate = 0.08

    # @type_definition(int)
    def calcTaxPrice(self, price):
        print("TaxEntity")
        p = self.__taxRate + 1
        return round(price * p)

class TaxCalculateUseCaseInputData:
    __price: int = 0

    # @type_definition(int)
    def __init__(self, price):
        print("TaxCalculateUseCaseInputData")
        self.__price = price

    def getPrice(self):
        return self.__price

class TaxCalculateUseCaseOutputData:
    __price: int = 0

    # @type_definition(int)
    def __init__(self, price):
        print("TaxCalculateUseCaseOutputData")
        self.__price = price

    def getPrice(self):
        return self.__price

class ITaxCalculateUseCase(metaclass=ABCMeta):
    @abstractmethod
    def calcTaxPrice(self, inputData: TaxCalculateUseCaseInputData):
        print("ITaxCalculateUseCase")
        pass

class TaxCalculateUseCase(ITaxCalculateUseCase):
    def calcTaxPrice(self, inputData: TaxCalculateUseCaseInputData):
        print("TaxCalculateUseCase")
        taxEntity = TaxEntity()
        price = inputData.getPrice()
        taxPrice = taxEntity.calcTaxPrice(price)
        return TaxCalculateUseCaseOutputData(taxPrice)


class ITaxCalculateView(metaclass=ABCMeta):
    @abstractmethod
    def display(self, outputData: TaxCalculateUseCaseOutputData):
        print("ITaxCalculateView")
        pass

class TaxCalculateView(ITaxCalculateView):
    def display(self, outputData: TaxCalculateUseCaseOutputData):
        print("TaxCalculateView")
        taxPrice = outputData.getPrice()
        print('taxPrice:%d ' % taxPrice)

class TaxCalculateController():
    # __useCase: iTaxCalculateUseCase = 
    def __init__(self, useCase: ITaxCalculateUseCase, view: ITaxCalculateView):
        self.useCase = useCase
        self.view = view

    def calcTaxPrice(self, priceText: str):
        price = int(priceText)
        inputData = TaxCalculateUseCaseInputData(price)
        outputData = self.useCase.calcTaxPrice(inputData)
        self.view.display(outputData)

for line in iter(sys.stdin.readline, ""):
    useCase = TaxCalculateUseCase()
    view = TaxCalculateView()
    controller = TaxCalculateController(useCase, view)
    controller.calcTaxPrice(line)

File: workshop/20190223-clean-architecture/ts-sample/sample-04.ts
import * as readline from 'readline';
import * as process from 'process';

class TaxEntity {
    private readonly taxRate: number = 0.08;

    public calcTaxPrice(price: number): number {
        const p = this.taxRate + 1;

        return Math.round(price * p);
    }
}

class TaxCalculateUseCaseInputData {
    public readonly price: number;

    constructor(price: number) {
        this.price = price;
    }
}

class TaxCalculateUseCaseOutputData {
    public readonly price: number;

    constructor(price: number) {
        this.price = price;
    }
}

interface ITaxCalculateUseCase {
    calcTaxPrice(inputData: TaxCalculateUseCaseInputData): TaxCalculateUseCaseOutputData;
}

class TaxCalculateUseCase implements ITaxCalculateUseCase {
    public calcTaxPrice(inputData: TaxCalculateUseCaseInputData): TaxCalculateUseCaseOutputData {
        const taxEntity = new TaxEntity();
        const price = inputData.price;
        const taxPrice = taxEntity.calcTaxPrice(price);

        return new TaxCalculateUseCaseOutputData(taxPrice);
    }
}

interface ITaxCalculateView {
    display(outputData: TaxCalculateUseCaseOutputData): void;
}

class TaxCalculateView implements ITaxCalculateView {
    public display(outputData: TaxCalculateUseCaseOutputData): void {
        const taxPrice = outputData.price;
        console.log(`税込価格:${taxPrice}`);
    }
}

class TaxCalculateController {
    private readonly useCase: ITaxCalculateUseCase;
    private readonly view: ITaxCalculateView;

    public constructor(useCase: ITaxCalculateUseCase, view: ITaxCalculateView) {
        this.useCase = useCase;
        this.view = view;
    }

    public calcTaxPrice(priceText: string): void {
        const price = Number(priceText);
        const inputData = new TaxCalculateUseCaseInputData(price);
        const outputData = this.useCase.calcTaxPrice(inputData);
        this.view.display(outputData);
    }
}

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

rl.on('line', (input: string) => {
    const useCase = new TaxCalculateUseCase();
    const view = new TaxCalculateView();
    const controller = new TaxCalculateController(useCase, view);
    controller.calcTaxPrice(input);
});

